### ISSUE TITLE
the eval code cost too much time

### ISSUE THREAD

**Issue by Chenly1993** [cvlab-stonybrook/DewarpNet#22]
When I use the provided eval code for counting ssim, it seems to cost too much time for each image. Do you have some suggestions to speed up？

**Comment by sagniklp** [comment]
depends on the packages we have used. Never tried to speed up. If you can find a faster alternative please share, thanks!

**Comment by Chenly1993** [comment]
1.I have tested the given models on the 130 images and evaled by the code of DocUnet. But the result is very different from the paper，which I get is MS-SSIM 0.236477 and LD 27.232672. Maybe the models are not the best models in Table 2. Can you provide the best models for a fair comparation?

2.I try to use the models you have provided to train jointly. The val_loss is decreasing. But the testing output of the models by joint training seems not to be better than the models before. Is there any details about training jointly I should pay attention?

3.All data used in training process is which you given from img_1.zip to img_2.zip ?

**Comment by sagniklp** [comment]
Test on the images given here: https://drive.google.com/drive/folders/1aPfQHGrGxpuIbYLONydbSkGNygRX2z2P?usp=sharing These are the output of the best model, I'll share the best model in future. 

Joint training doesn't improve results a lot, you should try training separately first to the best possible model you can get.
For BM regression separate training use GT WC images.

**Comment by Chenly1993** [comment]
Thank you for your reply !

> For BM regression separate training use GT WC images.

Is the images refer to the generated warped images ? But the code of BM regression separate training seems to use the checkerboard images.

**Comment by sagniklp** [comment]
The checkerboard images are used for reconstruction loss. GT WC means ground-truth world coordinate images.

**Comment by Chenly1993** [comment]
> Test on the images given here: https://drive.google.com/drive/folders/1aPfQHGrGxpuIbYLONydbSkGNygRX2z2P?usp=sharing These are the output of the best model, I'll share the best model in future.
> 
> Joint training doesn't improve results a lot, you should try training separately first to the best possible model you can get.
> For BM regression separate training use GT WC images.

I use the pictures which is given as the best results, and the eval code of DocUnet, the matlab I used is 2017b on linux. but I just got MS-SSIM  0.24663224615384607 LD 23.662958653846136.

**Comment by fh2019ustc** [comment]
> > Test on the images given here: https://drive.google.com/drive/folders/1aPfQHGrGxpuIbYLONydbSkGNygRX2z2P?usp=sharing These are the output of the best model, I'll share the best model in future.
> > Joint training doesn't improve results a lot, you should try training separately first to the best possible model you can get.
> > For BM regression separate training use GT WC images.
> 
> I use the pictures which is given as the best results, and the eval code of DocUnet, the matlab I used is 2017b on linux. but I just got MS-SSIM 0.24663224615384607 LD 23.662958653846136.

Hi, I used the Matlab 2019A on windows using the given Rectified Images,  and I get the exact result in the paper.

**Comment by Chenly1993** [comment]
> > > Test on the images given here: https://drive.google.com/drive/folders/1aPfQHGrGxpuIbYLONydbSkGNygRX2z2P?usp=sharing These are the output of the best model, I'll share the best model in future.
> > > Joint training doesn't improve results a lot, you should try training separately first to the best possible model you can get.
> > > For BM regression separate training use GT WC images.
> > 
> > 
> > I use the pictures which is given as the best results, and the eval code of DocUnet, the matlab I used is 2017b on linux. but I just got MS-SSIM 0.24663224615384607 LD 23.662958653846136.
> 
> Hi, I used the Matlab 2019A on windows using the given Rectified Images, and I get the exact result in the paper.

I used the Matlab 2019A on windows and resize the images to (598,400),  then the ms_ssim = 0.3393161923076922 ,LD = 5.886523976923077. Can you tell me more details?

**Comment by Chenly1993** [comment]
I  resized the images to h*w=598400 and keep the aspect ratio. And transform the rgb image to gray then I gained the result sms_ssim = 0.4676 ,LD=8.8968. Am I right?

**Comment by sagniklp** [comment]
You should get the exact same result as paper. I think something is still missing.

**Comment by Chenly1993** [comment]
> You should get the exact same result as paper. I think something is still missing. Could you try with matlab 2018?

the images should be transformed to gray, is this right?

**Comment by Chenly1993** [comment]
`addpath(fullfile(pwd,'mexDenseSIFT'));
addpath(fullfile(pwd,'mexDiscreteFlow'));

file_path='./scan/';
unwarp_file_path='./dewarpnet-public/';


img_path_list=struct2cell(dir(strcat(file_path,'*.png')));
unwarp_img_path_list=struct2cell(dir(strcat(unwarp_file_path,'*.png')));

[k len]=size(unwarp_img_path_list);
fid=fopen('ssim_results_dewarpnet.txt','wb')

for j = 1:len
    unwarp_image_name = unwarp_img_path_list{1,j};
    part=strsplit(char(unwarp_image_name), '_');
    org_img_name = char(strcat(part(1),'.png'));
    other_img=imread(strcat(unwarp_file_path,unwarp_image_name));
    [oth_rows,oth_cols,aaaa]=size(other_img);
    %disp(strcat(file_path,org_img_name))
    org_image=imread(strcat(file_path,org_img_name));
    %[org_rows,org_cols,aaaa]=size(org_image);
    org_image = imresize(org_image,[oth_rows,oth_cols]);
    %[oth_rows,oth_cols,aaaa]=size(other_img);
    [org_rows,org_cols,aaaa]=size(org_image);
    f1 = sqrt(598400/(org_cols*org_rows));
    f2 = sqrt(598400/(oth_rows*oth_cols));
    org_image_1 = imresize(org_image,[int32(org_rows*f1),int32(org_cols*f1)]);
    other_img_1 = imresize(other_img,[int32(oth_rows*f2),int32(oth_cols*f2)]);
    org_image_1 = rgb2gray(org_image_1);
    other_img_1 = rgb2gray(other_img_1);
    [ssim, ls] = evalUnwarp(other_img_1,org_image_1);
    fprintf(fid,'%s %f %f\n',unwarp_image_name,ssim,ls);
    fprintf('%s %f %f\n',unwarp_image_name,ssim,ls);
end 
`
I used 2018b and get ms ssim=0.4683 ld = 9.0707. I show the code I used here, can you tell me what's the wrong?

**Comment by sagniklp** [comment]
This is my eval script hope it'll help. Let me know if you still don't get the desired result. 
```
files=dir(in_folder);
mean_ms=0.0;
mean_ld=0.0;
ms_all=[];
ld_all=[];
for i=1:65
    ref_img=rgb2gray(imread(char(ref_folder+string(i)+'.png')));
    for j=1:2
        fname=string(i)+'_'+string(j)+' copy.png';
        char(in_folder+string(fname))
        tst_img=rgb2gray(imread(char(in_folder+string(fname))));
        %imshow(tst_img)
        [rh,rw,~]=size(ref_img);
        [th,tw,tc]=size(tst_img);
        ref_img=imresize(ref_img,sqrt(598400/(rh*rw)),'bicubic');
        [rh,rw,rc]=size(ref_img);
        tst_img=imresize(tst_img,[rh rw],'bicubic');
        %imshow(ref_img)
        %imshow(tst_img)
        [ms, ld]=evalUnwarp(tst_img, ref_img);
        fprintf(fileID,'%s , %f , %f\n',fname,ms,ld);
        mean_ms=mean_ms+ms;
        mean_ld=mean_ld+ld;
    end
end

mean_ms=mean_ms/130
mean_ld=mean_ld/130

fprintf(fileID,'\n---- mean , %f , %f ----\n',mean_ms,mean_ld);
fclose(fileID);
```

**Comment by Chenly1993** [comment]
> This is my eval script hope it'll help. Let me know if you still don't get the desired result.
> 
> ```
> files=dir(in_folder);
> mean_ms=0.0;
> mean_ld=0.0;
> ms_all=[];
> ld_all=[];
> for i=1:65
>     ref_img=rgb2gray(imread(char(ref_folder+string(i)+'.png')));
>     for j=1:2
>         fname=string(i)+'_'+string(j)+' copy.png';
>         char(in_folder+string(fname))
>         tst_img=rgb2gray(imread(char(in_folder+string(fname))));
>         %imshow(tst_img)
>         [rh,rw,~]=size(ref_img);
>         [th,tw,tc]=size(tst_img);
>         ref_img=imresize(ref_img,sqrt(598400/(rh*rw)),'bicubic');
>         [rh,rw,rc]=size(ref_img);
>         tst_img=imresize(tst_img,[rh rw],'bicubic');
>         %imshow(ref_img)
>         %imshow(tst_img)
>         [ms, ld]=evalUnwarp(tst_img, ref_img);
>         fprintf(fileID,'%s , %f , %f\n',fname,ms,ld);
>         mean_ms=mean_ms+ms;
>         mean_ld=mean_ld+ld;
>     end
> end
> 
> mean_ms=mean_ms/130
> mean_ld=mean_ld/130
> 
> fprintf(fileID,'\n---- mean , %f , %f ----\n',mean_ms,mean_ld);
> fclose(fileID);
> ```

With your code, I get the results mean_ms = 0.4695, mean_ld=8.9585
